src/dsp_tools/commands/xmlupload/models/ingest.py

```diff
@@ -0,0 +1,91 @@
+
+
+class IngestClient(Protocol):
+    def ingest(self, shortcode: str, filepath: Path) -> IngestResponse:
+        """Uploads a file to the ingest server and returns the IngestResponse.
+
+        This function sends a POST request to the ingest server with the file to upload.
+        It will retry the request 6 times in case of a connection, timeout or internal server error.
+
+        After all retry attempts are exhausted it will raise exceptions if the upload failed.
+        The http status code is also checked and if it is not 200, a PermanentConnectionError is raised.
+
+        Args:
+            shortcode: The shortcode of the project to ingest to.
+            filepath: Path to the file to ingest, could be either absolute or relative.
+
+        Raises:
+            BadCredentialsError: If the credentials are invalid.
+            PermanentConnectionError: If the connection fails.
+
+        """
+
+
+@dataclass(frozen=True)
+class DspIngestClient(IngestClient):
+    dsp_ingest_url: str
+    token: str
+
+    @staticmethod
+    def _retry_session(retries: int, session: Optional[Session] = None, backoff_factor: float = 0.3) -> Session:
+        session = session or requests.Session()
+        retry = Retry(
+            total=retries,
+            read=retries,
+            connect=retries,
+            backoff_factor=backoff_factor,
+            allowed_methods=None,  # means all methods
+            status_forcelist=[500],
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+        session.mount("http://", adapter)
+        session.mount("https://", adapter)
+        return session
+
+    def ingest(self, shortcode: str, filepath: Path) -> IngestResponse:
+        s = DspIngestClient._retry_session(retries=6)
+        url = f"{self.dsp_ingest_url}/projects/{shortcode}/assets/ingest/{filepath.name}"
+        err = f"Failed to ingest {filepath} to '{url}'."
+        try:
+            with open(filepath, "rb") as binary_io:
+                res = s.post(
+                    url=url,
+                    headers={"Authorization": f"Bearer {self.token}", "Content-Type": "application/octet-stream"},
+                    data=binary_io,
+                    timeout=60,
+                )
+                if res.status_code == 200:
+                    return IngestResponse(internal_filename=res.json()["internalFilename"])
+                elif res.status_code == 401:
+                    raise BadCredentialsError("Bad credentials")
+                else:
+                    user_msg = f"{err} See logs for more details: {logger_savepath}"
+                    print(user_msg)
+                    log_msg = f"{err}. Response status code {res.status_code} '{res.json()}'"
+                    logger.error(log_msg)
+                    raise PermanentConnectionError(log_msg)
+        except FileNotFoundError as e:
+            logger.error(e)
+            raise UserError(f"{err}. File {filepath} not found.")
```

Comment:
It is not a connection problem at this point. If the file is defined in xml and not present on the specified path this is IMHO a user error.

Reply:
I agree that it is not a "permanent connection error" if a file was specified that does not exist. So how about adding a `NonExistentFileError` which has the same effect of not aborting the upload, but recording the issue to report it to the user.

And on a more fundamental note: If this check was done before starting the upload, then it would be ok not to have to continue an upload that clearly isn't valid. (If there is no File uploaded, the resource can not be created; and then all resources having links to this resource will fail subsequently, etc.)

Reply:
I agree with Balduin's input, maybe as part of the restructuring of the classes we could also analyse which information we could implement as a fast fail and move that to the beginning.
One problem if we move this check to the beginning is though that if we do a `resume-xmlupload` then we cannot guarantee that we are still in the correct directory and the files were not changed... But I think this needs to be part of a larger discussion.

Reply:
I quickly checked, and the existence of the files is already checked at the very beginning in `_check_if_bitstreams_exist()`. So in fact, you should just remove the `except FileNotFoundError` block. It's not necessary here.

Reply:
done in b8317266d20bd312081e816f048b6954a3247e98